fix: solve login issue #95

jzelAdmin2006 opened the pull request:
Hi Trung

I came across this repo by chance and was enthusiastic about the idea. Unfortunately, when I tried it out, it didn't work and I ended up on this error page (it's German, because this is my browser language as a Swiss).

![image](https://github.com/trungvose/angular-spotify/assets/112761658/d16643d1-d8bd-4832-b6e8-c55bff9f6704)

The error apparently only occurs if you are already logged in to Spotify and only want to grant access to the app instead of logging in additionally. This is because Spotify does not accept plus signs that are coded with “%2520” (perhaps they used to accept it that way). However, everything seems to be OK if you don't encode the whole thing and just leave the '+' characters as they are.

![image](https://github.com/trungvose/angular-spotify/assets/112761658/728d488f-d226-45d7-9dd3-0f5388734169)

This seems to correspond with the issue #95. Thus, I have also left a comment there.

I hope I have been able to make a small contribution and, as a rookie, have not drawn a completely wrong conclusion 😅. If so, I have a question if you can give me a pro tip on how to get the Starstruck Achievment as easy as possible but legit (e.g. for my school project TouGH-Vault). I doubt that I have earned any form of endorsement, because that would probably be the fastest way, but prove me wrong 🙂.

All the best
JZEL

trungvose commented:
@jzelAdmin2006 Thanks for the PR! However, I'm quite confused because the scope doesn't contain any plus sign as you mentioned, and personally, it works for me. So maybe there's something else that I'm not aware of?

https://github.com/trungvose/angular-spotify/blob/86e28c1300bcea1f7c729d8e3092f211a0e6d215/libs/web/auth/data-access/src/lib/models/spotify-authorize.ts#L6

A sample link looks like

> https://accounts.spotify.com/authorize?client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=http%3A%2F%2Flocalhost%3A4200%2F&scope=user-read-recently-played%2520user-top-read%2520user-read-playback-position%2520user-read-playback-state%2520user-modify-playback-state%2520user-read-currently-playing%2520streaming%2520playlist-modify-public%2520playlist-modify-private%2520playlist-read-private%2520playlist-read-collaborative%2520user-library-modify%2520user-library-read%2520user-read-email%2520user-read-private&response_type=token

That's how the scope was intend on Spotify document

![image](https://github.com/trungvose/angular-spotify/assets/6767322/16d1396f-e98d-4cee-8bea-86298421fc17)

https://developer.spotify.com/documentation/web-playback-sdk/howtos/web-app-player

jzelAdmin2006 commented:
Hi Trungvose

Thank you for your response. I am afraid that you didn't have the same preconditions as me, since I realised that it can somehow recover itself with redirections, for example if you are **already signed in to Spotify**, or if you have **already activated the app on your account** (which I unfortunately didn't mention explicitly, sorry for that 🙂).
Regarding the plus signs, the client does not use any, as it is currently on "https://spotify.trungk18.com". As you wrote, there are these separators with "%2520". If you are either not yet logged in to Spotify or have already activated Angular-Spotify on your account before, the Spotify API doesn't actually process the parameter with these incorrect separators and just redirects you. With my change, a '+' sign is used instead of this "%2520", which also leads to a correct separation of the scopes according to this screenshot and is then also accepted by Spotify.

![image](https://github.com/trungvose/angular-spotify/assets/112761658/29f56c1f-2153-45dd-8382-e0b9d229c30b)

The version which is on "https://spotify.trungk18.com" delivers this request which leads to the error (if all the mentioned preconditions are fulfilled) (this is Bash CURL, of course I removed all the potentially sensitive stuff):

curl 'https://accounts.spotify.com/de/authorize/accept?ajax_redirect=1' \
  -H 'accept: */*' \
  -H 'accept-language: de-CH,de-DE;q=0.9,de;q=0.8,en-US;q=0.7,en;q=0.6' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'origin: https://accounts.spotify.com' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://accounts.spotify.com/de/authorize?client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=https%3A%2F%2Fspotify.trungk18.com%2F&scope=user-read-recently-played%2520user-top-read%2520user-read-playback-position%2520user-read-playback-state%2520user-modify-playback-state%2520user-read-currently-playing%2520streaming%2520playlist-modify-public%2520playlist-modify-private%2520playlist-read-private%2520playlist-read-collaborative%2520user-library-modify%2520user-library-read%2520user-read-email%2520user-read-private&response_type=token' \
  -H 'sec-ch-ua: "Chromium";v="125", "Not.A/Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?1' \
  -H 'sec-ch-ua-platform: "Android"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Mobile Safari/537.36' \
  --data-raw 'request=&client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=https%3A%2F%2Fspotify.trungk18.com%2F&scope=user-read-recently-played%2520user-top-read%2520user-read-playback-position%2520user-read-playback-state%2520user-modify-playback-state%2520user-read-currently-playing%2520streaming%2520playlist-modify-public%2520playlist-modify-private%2520playlist-read-private%2520playlist-read-collaborative%2520user-library-modify%2520user-library-read%2520user-read-email%2520user-read-private

My version, however, leads to this request, which works:

curl 'https://accounts.spotify.com/de/authorize/accept?ajax_redirect=1' \
  -H 'accept: */*' \
  -H 'accept-language: de-CH,de-DE;q=0.9,de;q=0.8,en-US;q=0.7,en;q=0.6' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'origin: https://accounts.spotify.com' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://accounts.spotify.com/de/authorize?client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=http%3A%2F%2Flocalhost%3A4200%2F&scope=user-read-recently-played+user-top-read+user-read-playback-position+user-read-playback-state+user-modify-playback-state+user-read-currently-playing+streaming+playlist-modify-public+playlist-modify-private+playlist-read-private+playlist-read-collaborative+user-library-modify+user-library-read+user-read-email+user-read-private&response_type=token' \
  -H 'sec-ch-ua: "Chromium";v="125", "Not.A/Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?1' \
  -H 'sec-ch-ua-platform: "Android"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Mobile Safari/537.36' \
  --data-raw 'request=&client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=http%3A%2F%2Flocalhost%3A4200%2F&scope=user-read-recently-played+user-top-read+user-read-playback-position+user-read-playback-state+user-modify-playback-state+user-read-currently-playing+streaming+playlist-modify-public+playlist-modify-private+playlist-read-private+playlist-read-collaborative+user-library-modify+user-library-read+user-read-email+user-read-private

Here is a step-by-step video where I reproduce the error in a fresh browser without history:

https://github.com/trungvose/angular-spotify/assets/112761658/2192d747-6fd0-4802-8177-bfdf64026864

Basically, I was making sure Angular-Spotify isn't activated on my account, then I log in to Spotify and I use the link "https://accounts.spotify.com/authorize?client_id=d06c09470bb646ebb33f27616fb151fb&redirect_uri=https%3A%2F%2Fspotify.trungk18.com%2F&scope=user-read-recently-played%2520user-top-read%2520user-read-playback-position%2520user-read-playback-state%2520user-modify-playback-state%2520user-read-currently-playing%2520streaming%2520playlist-modify-public%2520playlist-modify-private%2520playlist-read-private%2520playlist-read-collaborative%2520user-library-modify%2520user-library-read%2520user-read-email%2520user-read-private&response_type=token", which is the same that you provided me; the only difference is the URL that I changed from "http://localhost:4200" to "https://spotify.trungk18.com". If I'd typed in "https://spotify.trungk18.com", I would have been redirected to this.

I hope the issue is clear now and you are able to retrace my thoughts. Please let me know in case you need anything further.

All the best
JZEL

SegoCode commented:
Hey there, 

Just to remark that the analysis and change proposed by @jzelAdmin2006 is correct. Same problem from my side, I edited the request and sanitized the scope body, relaunched and get accepted and authorized request, now I can access the app

![2024-06-28_19-17](https://github.com/trungvose/angular-spotify/assets/35817798/2f134b05-8520-4e3a-bf64-9b4d9e54dba1)

trungvose commented:
Hi both, Sorry I overlooked the comment. I've been really busy recently. I managed to reproduce the issue pointed out by @jzelAdmin2006's video. Basically, if I remove Angular Spotify and try to open it for the first time, the issue appears. I believe they changed the API because this didn't happen three years ago. (well, times flies 😂)

Thanks so much for your effort debugging and fixing the app, @jzelAdmin2006 and @SegoCode! I'll merge the code and redeploy.

https://github.com/trungvose/angular-spotify/assets/6767322/6125da63-4185-465d-a582-db2d92f5724b
